### ISSUE TITLE
WriteFlusher debug of onFail does not log stacktraces

### ISSUE THREAD

**Issue by Maaartinus** [eclipse/jetty.project#1696]
TL;DR The bug is in `JettyAwareLogger.log(Marker marker, int level, String msg, Object[] argArray, Throwable t)` line 620.

This is a trivial bug causing exceptions not being logged. I'm using
```
org.eclipse.jetty.http2:http2-server:9.4.0.v20161208
org.eclipse.jetty:jetty-servlets:9.4.0.v20161208
org.slf4j:slf4j-api:1.7.25
ch.qos.logback:logback-classic:1.2.3
```
The intent is obviously to log the cause:
```
org.eclipse.jetty.io.WriteFlusher
public boolean onFail(Throwable cause)
LOG.debug("ignored: {} {}", this, cause);
cause = java.nio.channels.ClosedChannelException
```
But it's absent from the logfile. 

It gets passed further to
```
org.eclipse.jetty.util.log.Slf4jLog
public void debug(String msg, Object... args)
_logger.debug(msg, args);
args = [WriteFlusher@38211e27{IDLE}->null, java.nio.channels.ClosedChannelException]
```
and
```
org.eclipse.jetty.util.log.JettyAwareLogger
public void debug(String format, Object[] argArray)
log(null, DEBUG, format, argArray, null);
argArray = [WriteFlusher@38211e27{IDLE}->null, java.nio.channels.ClosedChannelException]
```
and
```
org.eclipse.jetty.util.log.JettyAwareLogger
public void debug(String format, Object[] argArray)
FormattingTuple ft = MessageFormatter.arrayFormat(msg, argArray);
argArray = [WriteFlusher@38211e27{IDLE}->null, java.nio.channels.ClosedChannelException]
```
Now, the `FormattingTuple` contains the exception, but it's stored separately, rather then formatted in the message. The next line
```
_logger.log(marker, FQCN, level, ft.getMessage(), null, t)
```
uses `t`, which is `null`. So the exception does not get logged.

Something like `t != null ? t : ft.getThrowable()` would solve it.

**Comment by joakime** [comment]
The line: https://github.com/eclipse/jetty.project/blob/jetty-9.4.0.v20161207/jetty-io/src/main/java/org/eclipse/jetty/io/WriteFlusher.java#L482

``` java
LOG.debug("ignored: {} {}", this, cause);
```

does not log a stacktrace.

It would have to be:

``` java
LOG.debug("ignored: "+ this, cause);
```

to have a stacktrace be logged.

It has nothing to do with slf4j or logback.

**Comment by joakime** [comment]
@sbordet up to you, do you want the stacktraces (or not) to show up in the debug log?

**Comment by Maaartinus** [comment]
> It has nothing to do with slf4j or logback.

It does as `JettyAwareLogger implements org.slf4j.Logger` just like `ch.qos.logback.classic.Logger` does and they behave differently in this case. The latter *does log* the exception when called as `logger.debug("ignored: {} {}", this, cause)`.

It produces something like "WriteFlusher@38211e27{IDLE}->null {}" followed by the exception (with stacktrace). The `{}` come form the special treatment of throwables: When the last argument is a throwable, it gets *always* shown and needs no `{}`. It's undocumented, but it works like I wrote.

It's a strange behavior, but IMHO better than silently dropping the exception (no idea what other loggers do). The overload `debug(String, Throwable)` is just a special case of this rule.

To summarize, with logback all these statements do log the exception with stacktrace:
```
logger.debug("string+object", (Object) e);
logger.debug("string+exception", e);
logger.debug("single-brace-pair+two-args: {}", o1, e);
logger.debug("two-brace-pairs+varargs: {} {}", o1, o2, e);
```

IMHO fully dropping the exception is a bug (at least something should be shown).

**Comment by sbordet** [comment]
Fixed as suggested by @joakime.
